### ISSUE TITLE
Simpler mechanism to enqueue tasks

### ISSUE THREAD

**Issue by slinkydeveloper** [rschmukler/agenda#367]
Hi,
For my application i need:
- Job scheduling **but**
- Flexibility to create job of various type
- **And** define job tasks when i want

So i think that the idea of define a task and then schedule this task is a bit complex because i have to first define all possible tasks at the start of agenda scheduler and **then** i can schedule it with datas.
So i write some code around agenda scheduler to create a simpler mechanism of task scheduling:
- The first step is to define a class (called for example `CoreEventQueue`) that define agenda scheduler, define the `event-resolve` agenda task and provide `registerEvent` and `removeEvent` functions:
```javascript
class CoreEventQueue {
    constructor() {
        this.agenda = new Agenda({db: {address: Config.MONGODB_CONNECTION_STRING + '/event-scheduler'}});
        this.agenda.on('ready', () => {
            this.agenda.start();
        });

        this.agenda.define('event-resolve', (job, done) => {
            job.attrs.data.event.resolveEvent(done);
        });
    }

    registerEvent(e, at) { // e it's an Event object
            let jobData = {event: e, eventName: e.getEventName()}; // eventName it's a unique event id that can identify a unique event
            if (at) {
                this.agenda.schedule(at, 'event-resolve', jobData);
            } else {
                this.agenda.now('event-resolve', jobData);
            }
        }
    }

    removeEvent(eventName, cb) {
        this.agenda.cancel({name: 'event-resolve', "data.eventName": eventName}, cb);
    }
}
```
- Next i define the `Event` class, the class that **you must subclass** if you want to create an event:
```javascript
var CoreEventQueue = require('../CoreEventQueue.class.js');

class Event {
    constructor() {
        if (new.target === Event)
            throw new TypeError("Cannot instance Event class. You need to instance a subclass of Event class");
    }

    getEventName() {
        return this.eventName;
    }

    resolveEvent(cb) {
        cb();
    }

    registerToEventQueue(at) {
        CoreEventQueue.registerEvent(this, at);
    }
};
```
- Now i can define the events subclassing the `Event` class. This way is more flexible to create an _event_type_:
```javascript
class YetAnotherEvent extends Event {
    constructor(someData) {
        super();
        this.eventName = 'YetAnotherEvent-' + uuidGenerate();

        this.someData = someData;
    }

    resolveEvent(cb) { // the function that will be executed by agenda scheduler
        doSomething(this.someData, cb);
    }
};
```
- Now every time that i need to schedule a new YetAnotherEvent i execute this simple function:
```javascript
new YetAnotherEvent(someData).registerToEventQueue(dateToExecute);
```

I need this mechanism for my express application because it's useful to run some tasks  (for example send a push notification) after a user request to my API or to schedule jobs (for example a removal of user at midnight of the same day) after an user request. This way it's simpler because i don't have to first define all tasks, but i can define a task when i want and then execute it.
And now comes the truth: when i try to schedule the task, it doesn't work because the jobData object is too big. I understand the problem to serialize functions that some times are too big, but it can be really useful to use an events mechanism like this, ES6 style! And also give more flexibility, because wherever part of code you can define a task type and schedule that.

In conclusion, there's a way to get this mechanism working or i have to rewrite my code? Or I think that you can implement this events mechanism (eventually in a fork) because it's really useful!

**Comment by OmgImAlexis** [comment]
PING: @agenda/maintainers

**Comment by slinkydeveloper** [comment]
Hi! Finally I found how to solve this kind of problem and i've implemented it inside my project, if you want we can discuss on how to implement it directly inside agenda.
